[stable/cockroachdb] Complete refactor and capabilities extension

tyranron opened the pull request:
This is a hard backwards-incompatible rework of the `stable/cockroachdb` chart.

Fixes #18399  
Possibly fixes #18402 (at least in all tested cases by me it has passed)


## Motivation

The current `2.x.x` version of `stable/cockroachdb` is hard to reuse in a variety of situations.

Specifically, it bounds you to:
- use only PersistentVolumes for storing data
- use only public-accessible Docker images
- doesn't allow to tweak affinity rules and tolerations

All this made it hard for us to reuse Helm chart both for production/staging/local-dev environments and even numerous E2E tests.


## BC Breaks

The version of the chart with this PR is bumped up to `3.0.0`, reflecting the following backwards-compatibility breaks:
- Values of the chart are fully renamed to follow `camelCase` style (instead of `UpperCase`) widely used in Helm chart's ecosystem (fixes #18399).
    - Additionally, values passed as CLI args to `cockroach` follow the `kebab-case` style to represent a direct mapping from values to these CLI args and omit possible misunderstandings.
- Labels `chart`, `heritage`, `release` and `component` are replaced with more idiomatic `helm.sh/chart`, `app.kubernetes.io/name`, `app.kubernetes.io/instance`, `app.kubernetes.io/managed-by` and `app.kubernetes.io/component`.
- Explicitly differentiate the listening ports in Pods (`conf.port`/`conf.http-port`) and in Services (`service.ports`).
- Remove the `Name` value. Now, the name is idiomatically consists of release name + chart name and can explicitly overwritten with `nameOverride` and `fullnameOverride` values.
- Remove the `Component` value. Now the components of the chart may be differentiated with specifying custom labels for them, if required.


## Added

- Support `nameOverride` and `fullnameOverride` values.
- Ability to specify credentials for pulling used Docker images from private registries.
- Ability to label and annotate almost any chart's resource either dependently and independently (see `values.yml` for details).
- Ability to configure `affinity` rules, `tolerations` taints and `nodeSelector` both for StatefulSet and initialization Job.
- Ability to persist data in:
    - `hostPath` for those who doesn't want to poke (or is not able to) with PersistentVolumes;
    - `emptyDir` which is convenient for temporary snapshot-based deployments (when the container already holds its data and persistence is unnecessary);
    - `persistentVolume` which is still the default one and now supports additional labels and annotations.
- Ability to specify custom ingress rules for NetworkPolicy in addition to existing ones.
- Ability to reuse entrypoint of the used Docker image (previously it was hardly overwritten), which adds the ability to run custom logic before the `cockroach` starts (in our case, for example, this is the data restoration from a snapshot, widely used for numerous E2E tests).
- Ability to deploy in standalone mode (`cockroach start-single-node`) without clustering.
- Ability to configure `logtostderr` verbosity.
- Ability to configure `cluster-name` and `disable-cluster-name-verification`.
- Ability to configure more CLI args as `conf`chart values. While still allowing specify custom CLI extra args (via `statefulset.args`).


## Additionally

From the chart description I've noticed that there are some chart instructions outside, specifically on CockroachDB's site:
> https://www.cockroachlabs.com/docs/v19.2/orchestrate-cockroachdb-with-kubernetes-insecure.html

I'll create PR and update those once (and if) this PR is merged.


## Checklist

- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped (to `3.0.0`)
- [x] All the variables are documented in the `README.md`
- [x] `README.md` info is updated and upgrading from previous versions is described.
- [x] Title of the PR starts with chart name (e.g. `[stable/cockroachdb]`)

DuskEagle commented on stable/cockroachdb/templates/_helpers.tpl:
```diff
@@ -1,32 +1,65 @@
+{{/* vim: set filetype=mustache: */}}
```
Can we remove the vim comment here?

DuskEagle commented on stable/cockroachdb/templates/job.init.yaml:
```diff
@@ -0,0 +1,118 @@
+{{- if gt (.Values.statefulset.replicas | int64) 1 }}
```
I think this should also check if `{{- if .Values.conf.join }}` is set. If a `join` flag is passed to the `cockroach start` command in the statefulset, then this job is required, even when only running one CRDB pod.

tyranron committed:
[stable/cockroachdb] Complete refactor and capabilities extension

- refactor chart values to be more idiomatic and document them more precisely
- rename k8s labels to namespaced ones
- refactor template names and make them more readable
- support 'nameOverride' and 'fullnameOverride' values
- make StatefulSet 'replicas' configurable
- make affinity, tolerations and nodeSelector configurable for StatefulSet and init Job
- make labels and annotations configurable for Pods of StatefulSet and init Job
- make init Job idempotent for already initialized clusters
- add ability to persist data in hostPath and emptyDir
- add ability to configure NetworkPolicy with custom ingress rules
- add ability to specify credentials for private registries of Docker images
- support running in standalone mode (start-single-node)
- provide more configuration parameters in chart values
- reuse entrypoint of the used Docker image instead of overwriting it
- bump chart's version to 3.0.0
- update README and add upgrading instructions for 3.0.0

Signed-off-by: tyranron <tyranron@gmail.com>
